cpp/src/quantiles/tdigest/tdigest.cu

```diff
@@ -0,0 +1,230 @@
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/binary_search.h>
+
+namespace cudf {
+namespace detail {
+
+namespace {
+
+__device__ double lerp(double a, double b, double t) { return a + t * (b - a); }
+
+// kernel for computing percentiles on input tdigest (mean, weight) centroid data.
+__global__ void compute_percentiles_kernel(offset_type const* tdigest_offsets,
+                                           size_type num_tdigests,
+                                           double const* percentages,
+                                           size_type num_percentages,
+                                           double const* mean_,
+                                           double const* weight_,
+                                           double const* cumulative_weight_,
+                                           double* output)
+{
+  int const tid = threadIdx.x + blockIdx.x * blockDim.x;
+
+  auto const tdigest_index = tid / num_percentages;
+  if (tdigest_index >= num_tdigests) { return; }
+  double const percentage = percentages[tid % num_percentages];
+
+  // size of the digest we're querying
+  auto const tdigest_size = tdigest_offsets[tdigest_index + 1] - tdigest_offsets[tdigest_index];
+  double const* cumulative_weight = cumulative_weight_ + tdigest_offsets[tdigest_index];
+
+  // means and weights for the tdigest
+  double const* mean   = mean_ + tdigest_offsets[tdigest_index];
+  double const* weight = weight_ + tdigest_offsets[tdigest_index];
+
+  // if we're at either end
+  double const total_weight = cumulative_weight[tdigest_size - 1];
+  double const cluster_q    = percentage * total_weight;
+  if (cluster_q <= 1) {
+    output[tid] = mean[0];
+    return;
+  } else if (cluster_q >= total_weight - 1) {
+    output[tid] = mean[tdigest_size - 1];
+    return;
+  }
+
+  // otherwise find the centroid we're in and interpolate
+  size_type const centroid_index = static_cast<size_type>(
+    thrust::lower_bound(
+      thrust::seq, cumulative_weight, cumulative_weight + tdigest_size, cluster_q) -
+    cumulative_weight);
+
+  double diff = cluster_q + weight[centroid_index] / 2 - cumulative_weight[centroid_index];
+  if (weight[centroid_index] == 1 && std::abs(diff) < 0.5) {
+    output[tid] = mean[centroid_index];
+    return;
+  }
+  size_type left_index  = centroid_index;
+  size_type right_index = centroid_index;
+  if (diff > 0) {
+    if (right_index == tdigest_size - 1) {
+      output[tid] =
+        lerp(mean[right_index], mean[tdigest_size - 1], diff / (weight[right_index] / 2));
+      return;
+    }
+    right_index++;
+  } else {
+    if (left_index == 0) {
+      output[tid] = lerp(mean[0], mean[left_index], diff / (weight[left_index] / 2));
+      return;
+    }
+    left_index--;
+    diff += weight[left_index] / 2 + weight[right_index] / 2;
+  }
+
+  diff /= (weight[left_index] / 2 + weight[right_index] / 2);
+  output[tid] = lerp(mean[left_index], mean[right_index], diff);
+}
```

Comment:
There's a lot of Initialize-Then-Modify going on here with `diff` as well as `output[tid]` with several return points. I'd suggest trying to see if either or both of these things can be const-calculated